iroha_actor/src/broker.rs

```diff
@@ -53,65 +13,132 @@ pub async fn issue_send<M: BrokerMessage + Send + Sync>(m: M) {
-///     Actor2.init().start().await;
-///     Actor1.init().start().await;
+///     let broker = Broker::new();
+///     Actor2(broker.clone()).start().await;
+///     Actor1(broker).start().await;
 /// })
 /// ```
-#[async_trait::async_trait]
-pub trait BrokerActor: Actor {
-    /// Subscribe actor to specific message type
-    fn subscribe<M: BrokerMessage>(&self, ctx: &mut Context<Self>)
-    where
-        Self: Handler<M>,
-    {
-        let mut entry = message_entry(TypeId::of::<M>()).or_insert_with(|| Vec::with_capacity(1));
+use std::any::{Any, TypeId};
+use std::sync::Arc;
+
+use dashmap::mapref::entry::Entry;
+use dashmap::DashMap;
+use futures::future;
+
+use super::*;
+
+type TypeMap<V> = DashMap<TypeId, V>;
+type BrokerRecipient = Box<dyn Any + Sync + Send + 'static>;
+
+/// Broker type. Can be cloned and shared between many actors.
+///
+/// TODO: There might be several actors of one type. We should handle this case!
+#[derive(Debug)]
+pub struct Broker(Arc<TypeMap<Vec<(TypeId, BrokerRecipient)>>>);
+
+impl Clone for Broker {
+    fn clone(&self) -> Self {
+        Self(Arc::clone(&self.0))
+    }
+}
+
+impl Default for Broker {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Broker {
+    /// Default constructor for broker
+    pub fn new() -> Self {
+        Self(Arc::new(DashMap::new()))
+    }
+
+    fn message_entry(&'_ self, id: TypeId) -> Entry<'_, TypeId, Vec<(TypeId, BrokerRecipient)>> {
+        self.0.entry(id)
+    }
+
+    /// Send message via broker
+    pub async fn issue_send<M: BrokerMessage + Send + Sync>(&self, m: M) {
+        let entry = if let Entry::Occupied(entry) = self.message_entry(TypeId::of::<M>()) {
+            entry
+        } else {
+            return;
+        };
+        let send = entry.get().iter().filter_map(|(_, recipient)| {
+            recipient
+                .downcast_ref::<Recipient<M>>()
+                .map(|recipient| recipient.send(m.clone()))
+        });
+        drop(future::join_all(send).await);
+    }
+
+    fn subscribe_recipient<M: BrokerMessage>(&self, recipient: Recipient<M>) {
+        let mut entry = self
+            .message_entry(TypeId::of::<M>())
+            .or_insert_with(|| Vec::with_capacity(1));
         if entry
             .iter()
-            .any(|(actor_id, _)| actor_id == &TypeId::of::<Self>())
+            .any(|(actor_id, _)| *actor_id == TypeId::of::<Self>())
         {
             return;
         }
-        entry.push((TypeId::of::<Self>(), Box::new(ctx.recipient::<M>())));
+        entry.push((TypeId::of::<Self>(), Box::new(recipient)));
+    }
+
+    /// Subscribe actor to specific message type
+    pub fn subscribe<M: BrokerMessage, A: Actor + ContextHandler<M>>(&self, ctx: &mut Context<A>) {
+        self.subscribe_recipient(ctx.recipient::<M>())
+    }
+
+    /// Subscribe with channel to specific message type
+    pub fn subscribe_also<M: BrokerMessage + Debug>(&self) -> mpsc::Receiver<M> {
```

Comment:
```suggestion
    pub fn subscribe_with_channel<M: BrokerMessage + Debug>(&self) -> mpsc::Receiver<M> {
```

Reply:
Idk, new name seems too long. Once you've seen signature you can remember it.

Reply:
Not very long and actually quite usual for Rust. This function is very confusing due to its name. It should definitely be changed. I think at least let's use this one. If you have other suggestions for the name we might consider.

The name of the function should be very descriptive, otherwise it is a pain to read the code.